src/target/xtensa/xtensa.c

```diff
@@ -975,176 +1054,179 @@ int xtensa_fetch_all_regs(struct target *target)
-}
-
-int xtensa_fetch_user_regs_u32(struct target *target)
-{
-	struct xtensa *xtensa = target_to_xtensa(target);
-	struct reg *reg_list = xtensa->core_cache->reg_list;
-	xtensa_reg_val_t cpenable = 0;
-	uint8_t regvals[XT_USER_REGS_NUM_MAX][sizeof(xtensa_reg_val_t)];
-	uint8_t dsrs[XT_USER_REGS_NUM_MAX][sizeof(xtensa_dsr_t)];
-	bool debug_dsrs = !xtensa->regs_fetched || LOG_LEVEL_IS(LOG_LVL_DEBUG);
-
-	assert(xtensa->core_config->user_regs_num < XT_USER_REGS_NUM_MAX && "Too many user regs configured!");
-	if (xtensa->core_config->coproc)
-		cpenable = xtensa_reg_get(target, XT_REG_IDX_CPENABLE);
-
-	for (unsigned int i = 0; i < xtensa->core_config->user_regs_num; i++) {
-		if (!xtensa_reg_is_readable(xtensa->core_config->user_regs[i].flags, cpenable))
-			continue;
-		xtensa_queue_exec_ins(xtensa, XT_INS_RUR(xtensa->core_config->user_regs[i].reg_num, XT_REG_A3));
-		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(XT_SR_DDR, XT_REG_A3));
-		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, regvals[i]);
-		if (debug_dsrs)
-			xtensa_queue_dbg_reg_read(xtensa, NARADR_DSR, dsrs[i]);
-	}
-	/* Ok, send the whole mess to the CPU. */
-	int res = jtag_execute_queue();
-	if (res != ERROR_OK) {
-		LOG_ERROR("Failed to fetch AR regs!");
-		return res;
-	}
-	xtensa_core_status_check(target);
-
-	if (debug_dsrs) {
-		/* DSR checking: follows order in which registers are requested. */
-		for (unsigned int i = 0; i < xtensa->core_config->user_regs_num; i++) {
-			if (!xtensa_reg_is_readable(xtensa->core_config->user_regs[i].flags, cpenable))
-				continue;
-			if (buf_get_u32(dsrs[i], 0, 32) & OCDDSR_EXECEXCEPTION) {
-				LOG_ERROR("Exception reading %s!", xtensa->core_config->user_regs[i].name);
-				return ERROR_FAIL;
+			if ((rlist[ridx].flags & XT_REGF_MASK) == XT_REGF_NOREAD) {
+				/* Report read-only registers all-zero but valid */
+				reg_list[i].valid = true;
+				xtensa_reg_set(target, i, 0);
+			} else {
+				reg_list[i].valid = false;
 			}
 		}
 	}
 
-	for (unsigned int i = 0; i < xtensa->core_config->user_regs_num; i++) {
-		if (xtensa_reg_is_readable(xtensa->core_config->user_regs[i].flags, cpenable)) {
-			buf_cpy(regvals[i], reg_list[XT_USR_REG_START + i].value, reg_list[XT_USR_REG_START + i].size);
-			reg_list[XT_USR_REG_START + i].valid = true;
-		} else {
-			reg_list[XT_USR_REG_START + i].valid = false;
-		}
-	}
-
-	/* We have used A3 as a scratch register and we will need to write that back. */
+	if (xtensa->core_config->windowed) {
+		/* We have used A3 as a scratch register.
+		 * Windowed configs: restore A3's AR (XT_REG_GENERAL) and and flag for write-back.
+		 */
+		enum xtensa_reg_id ar3_idx = xtensa_windowbase_offset_to_canonical(xtensa, XT_REG_IDX_A3, windowbase);
+		xtensa_reg_set(target, ar3_idx, a3);
+		xtensa_mark_register_dirty(xtensa, ar3_idx);
+
+		/* Reset scratch_ars[] on fetch.  .chrval tracks AR mapping and changes w/ window */
+		sprintf(xtensa->scratch_ars[XT_AR_SCRATCH_AR3].chrval, "ar%d", ar3_idx - XT_REG_IDX_AR0);
```

Comment:
Since GDB allows the user to update either A3 or AR3 (and A4 or AR4) as if they are fully independent registers, and the hardware maps them to the same underlying resource depending on windowbase, I keep track of both here in order to resolve cases where the user modifies one of them while the underlying register contains scratch data.